ts/query.go

```diff
@@ -0,0 +1,431 @@
+// iterator thus maintains a two-level index to point to a unique sample.
+type dataSpanIterator struct {
+	*dataSpan
+	offset    int32 // The normalized offset of the current sample within the dataSpan
+	dataIdx   int   // Index of InternalTimeSeriesData which contains current Sample
+	sampleIdx int   // Index of current Sample within InternalTimeSeriesData
+	valid     bool  // True if this iterator points to a valid Sample
+}
+
+// sample returns the InternalTimeSeriesSample value currently pointed to by
+// this iterator.
+func (dsi *dataSpanIterator) sample() *proto.InternalTimeSeriesSample {
+	if !dsi.valid {
+		return nil
+	}
+	return dsi.datas[dsi.dataIdx].Samples[dsi.sampleIdx]
+}
+
+// advance moves the iterator to point to the next Sample.
+func (dsi *dataSpanIterator) advance() {
+	if !dsi.valid {
+		return
+	}
+	data := dsi.datas[dsi.dataIdx]
+	switch {
+	case dsi.sampleIdx+1 < len(data.Samples):
+		dsi.sampleIdx++
+	case dsi.dataIdx+1 < len(dsi.datas):
+		dsi.dataIdx++
+		data = dsi.datas[dsi.dataIdx]
+		dsi.sampleIdx = 0
+	default:
+		dsi.valid = false
+	}
+	dsi.offset = data.offsetAt(dsi.sampleIdx)
+}
+
+// interpolatingIterator is used to iterate over offsets within a dataSpan. The
+// iterator can provide sample values for any offset, even if there is no actual
+// sample in the dataSpan at that offset.
+//
+// Values for missing offsets are computed using linear interpolation from the
+// nearest real samples preceding and following the missing offset.
+type interpolatingIterator struct {
+	offset   int32            // Current offset within dataSpan
+	nextReal dataSpanIterator // Next sample with an offset >= iterator's offset
+	prevReal dataSpanIterator // Prev sample with offset < iterator's offset
+}
+
+// advanceTo advances the iterator to the supplied offset.
+func (ii *interpolatingIterator) advanceTo(offset int32) {
+	ii.offset = offset
+	// Advance real iterators until nextReal has offset >= the interpolated
+	// offset.
+	for ii.nextReal.valid && ii.nextReal.offset < ii.offset {
+		ii.prevReal = ii.nextReal
+		ii.nextReal.advance()
+	}
+}
+
+// isValid returns true if this interpolatingIterator still points to valid data.
+func (ii *interpolatingIterator) isValid() bool {
+	return ii.nextReal.valid
+}
+
+// avg returns the average value at the current offset for this iterator.
+func (ii *interpolatingIterator) avg() float64 {
+	if !ii.isValid() {
+		return 0
+	}
+	if ii.nextReal.offset == ii.offset {
+		return ii.nextReal.sample().Average()
+	}
+	// Cannot interpolate if previous value is invalid.
+	if !ii.prevReal.valid {
+		return 0
+	}
+
+	// Linear interpolation of value at the current offset.
+	off := float64(ii.offset)
+	nextAvg := ii.nextReal.sample().Average()
+	nextOff := float64(ii.nextReal.offset)
+	prevAvg := ii.prevReal.sample().Average()
+	prevOff := float64(ii.prevReal.offset)
+	return prevAvg + (nextAvg-prevAvg)*(off-prevOff)/(nextOff-prevOff)
```

Comment:
Actually, it is not possible because of the way advanceTo() works; if two subsequent Samples had the same offset, advanceTo() would skip over the entire block.